<img src="../media/melbioinf_logo.png" width="350"> <img src="../media/PRIMARY_A_Vertical_Housed_RGB.png" width="150">

# Introduction to R for Biologists

Anticipated workshop duration when delivered to a group of participants is **3 hours**.  

For queries relating to this workshop, contact Melbourne Bioinformatics (bioinformatics-training@unimelb.edu.au).

## Overview

### Topic

* [ ] Genomics
* [ ] Transcriptomics
* [ ] Proteomics
* [ ] Metabolomics
* [x] Statistics and visualisation
* [ ] Structural Modelling
* [x] Basic skills


### Skill level

* [x] Beginner  
* [ ] Intermediate  
* [ ] Advanced  

This workshop is is aimed at bench biologists with no coding/programming skills.


### Description

*Ever wanted to visualise your own data? You can with R!*

This introduction to R and RStudio will provide beginners with experience with loading, manipulating and visualising biological data using the tidyverse collection of R packages. The example data used is publicly available RNA-seq data, therefore attendees will gain experience in the structure and appearance of RNA-seq data.
This workshop has been developed collaboratively by training specialists at Peter MacCallum Cancer Centre and Melbourne Bioinformatics.

Tools: [R](https://www.r-project.org/), [RStudio](https://rstudio.com/), [tidyverse](https://www.tidyverse.org/), [ggplot2](https://ggplot2.tidyverse.org/)

-------------------------------
## Learning Objectives

At the end of this introductory workshop, you will be able to:

*	Load tabular data into R.
*	Apply tidyverse functions to manipulate data in R.
*	Produce simple plots such boxplots using ggplot.
*	Understand and apply data faceting in ggplot.
*	Modify the aesthetics of a ggplot plot.

-------------------------------
## Requirements and preparation

**Attendees are required to bring their own laptop computers to face to face workshops.** <br/>

At least one week before the workshop, participants should install the software and data files below.  This should provide sufficient time for participants to liaise with their own IT support should they encounter any IT problems.  
We recommend you watch this [video](https://resources.rstudio.com/tidyverse/tidyverse-visualisation-manipulation-basics) to familiarise yourself with R, Rstudio and Tidyverse.


### Preparing your laptop prior to starting this workshop
1. Download and install [R](https://cran.r-project.org/) (free) on your computer.
2. Download and install [RStudio](https://rstudio.com/products/rstudio/download/#download) (free) on your computer.
3. Detailed instructions for installing R and RStudio are available [here](https://drive.google.com/file/d/1Bso4B7_ZxI_7q9K_TVnPoeQIQFK_9TSI/view).

### Required Software
* [R](https://software.broadinstitute.org/software/igv/download)
* [Rstudio](https://...)

### Required Data
* Data file can be downloaded from [here](https://github.com/melbournebioinformatics/r-intro-biologists/raw/master/data.zip)
* Hyper links for downloading the required data are also included in the workshop PDF, see 'Workshop material' below.

-------------------------------
## Author Information
Written by: Maria Doyle, Jessica Chung and Victoria Perreau  
This workshop has been developed collaboratively by training specialists at Peter MacCallum Cancer Centre and Melbourne Bioinformatics

**Created:** October 2019<br/>
**Last Reviewed:** October 2020

-------------------------------

## Workshop Material

Download the PDF with detailed workshop instructions [here](https://raw.githubusercontent.com/melbournebioinformatics/r-intro-biologists/master/intro_r_biologists.pdf). Hyperlinks to the required data are also included in the PDF.

-------------------------------
## Additional reading
* [A short intro to R and tidyverse](https://pmacdasci.github.io/r-intro-tidyverse/)
* [Topp 50 ggplot visualisations](http://r-statistics.co/Top50-Ggplot2-Visualizations-MasterList-R-Code.html)
* [R for Data Science](https://r4ds.had.co.nz/)
