# 10X single-cell RNA-seq analysis in R

## Overview

In this workshop, you will be learning how to analyse 10X Chromium single-cell RNA-seq profiles using R. This will include reading the count data into R, quality control, normalisation, dimensionality reduction, cell clustering and finding marker genes. The main part of the workflow uses the package. You will learn how to generate common plots for visualising single-cell data, such as t-SNE plots and heatmaps. This workshop is aimed at biologists interested in learning how to explore single-cell RNA-seq data generated by the 10X platform.

This workshop is presented by Dr Yunshun Chen from Walter Eliza Hall Institute of Medical Research (WEHI). Dr Chen is a statistical bioinformatician and a senior post-doc at WEHI Bioinformatics Division. He is one of the authors and the main maintainer of the edgeR package. He has extensive experience in RNA-seq gene expression and single-cell RNA-seq analyses.

## Requirements

The course is aimed at advanced PhD students, postdoctoral researchers and principal investigators. Some basic R knowledge is assumed – this is not an introduction to R course. If you are not familiar with the R statistical programming language it is compulsory that you work through an introductory R course before you attend this workshop.

This workshop will cover single-cell RNA-seq analysis and assumes you have some familiarity with the more common analysis of bulk RNA-seq data. If you have no experience in analysing bulk RNA-seq data, we strongly recommend you also attend our RNA-seq Differential Gene Expression analysis in R workshop.

Participants must bring a laptop with a Mac, Linux, or Windows operating system (not a tablet, Chromebook, etc.) that they have administrative privileges on. Bringing your laptop charger is advised. R and a few specific software packages should be installed in advance. The 10X data to be analysed, the gene annotation file and the gene signature RData file should also be downloaded prior to the workshop to allow us time to troubleshoot any problems you may have.

Downloads:

1. Download R from [https://cran.r-project.org/](https://cran.r-project.org/). The lastest version is recommended (version 3.6.0 for Windows and Mac OS X 10.11 and higher).
2. Install R packages by opening R and copying the following commands into your R console

```
if (!requireNamespace(“BiocManager”, quietly = TRUE))

    install.packages(“BiocManager”)

BiocManager::install(c(“scran”, “monocle”, “vcd”))
```

3. Download the 10X single-cell RNA-seq data from GEO (https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSM2510617). The three data files required for the workshop are: GSM2510617_P7-barcodes.tsv.gz, GSM2510617_P7-genes.tsv.gz and GSM2510617_P7-matrix.mtx.gz, of which the download links are available at the bottom of the web page. Unzip the three data files and store them in the working directory (or a sub-folder under the working directory).
4. Download the mouse gene annotation file from [http://bioinf.wehi.edu.au/edgeR/Mus_musculus.gene_info.gz](http://bioinf.wehi.edu.au/edgeR/Mus_musculus.gene_info.gz) and save it under the working directory. Note there is no need to unzip the file.
5. Download gene signatures from [http://bioinf.wehi.edu.au/edgeR/BulkSignatures.RData](http://bioinf.wehi.edu.au/edgeR/BulkSignatures.RData) and save it under the working directory.
6. Check downloads are correct by looking for any error messages when running the commands:

```
library(edgeR)
library(scater)
library(scran)
library(monocle)
library(vcd)
read.delim(“P7-genes.tsv”, nrow=5)
read.delim(“Mus_musculus.gene_info.gz”, nrow=5)
load(“BulkSignatures.RData”)
```

Important: If you have any trouble installing the software or packages, please contact us prior to the workshop.

## Workshop syllabus

This workshop will follow [this analysis guide](http://bioinf.wehi.edu.au/edgeR/10X.pdf).

## Additional single-cell analysis links

* An overview and comparison of protocols and analysis methods: [Tian, L. et al, Biorxiv, scRNA-seq mixology: towards better benchmarking of single cell RNA-seq protocols and analysis methods](https://www.biorxiv.org/content/biorxiv/early/2018/10/08/433102.full.pdf)

* Single cell RNA-seq data analysis:
    * [A course by CSC, Finland: Single cell data analysis](https://github.com/bishwaG/Single-cell-RNA-seq-data-analysis-2018)
    * [A course by the Hemberg Lab: scRNA-seq analysis](https://www.sanger.ac.uk/science/tools/scrna-seq-analysis-course)

* List of software packages for single-cell data analysis: [Davis, S. et al: Zenodo link](https://doi.org/10.5281/zenodo.1117762); [Github link](https://github.com/seandavi/awesome-single-cell)
